containers/Loans/Loans.tsx

```diff
@@ -56,133 +56,67 @@ function Container() {
 			setIsLoadingLoans(true);
-			const loanIndices = await Promise.all(Object.keys(loanStateContracts).map(getLoanIndices));
-			if (isMounted) {
-				loanIndices.forEach(({ type, minCRatio }) => {
-					setMinCRatios((cratios) => cratios.set(type, minCRatio.mul(1e2)));
-				});
-			}
-
-			const loans: Array<any> = await Promise.all(loanIndices.map(getLoans));
-			let activeLoans: Array<any> = [];
-
-			loans.forEach((a) => {
-				a.forEach(({ type, minCRatio, loan }: any) => {
-					if (!loan.amount.eq(0)) {
-						activeLoans.push({
-							loan,
-							type,
-							minCRatio,
-						});
-					}
-				});
-			});
-
-			activeLoans = await Promise.all(activeLoans.map(makeLoan));
-			activeLoans.sort((a, b) => {
-				if (a.id.gt(b.id)) return -1;
-				if (a.id.lt(b.id)) return 1;
-				return 0;
+			const openLoans = await getOpenLoans({
+				loanContracts,
+				loanStateContracts,
+				address,
+				network,
+				isL2,
 			});
-
 			if (isMounted) {
-				setLoans(activeLoans);
+				setLoans(openLoans);
 				setIsLoadingLoans(false);
 			}
 		};
 
-		const getLoanIndices = async (type: string) => {
-			const loanContract = loanContracts[type];
-			const loanStateContract = loanStateContracts[type];
-			const [n, minCRatio] = await Promise.all([
-				loanStateContract.getNumLoans(address),
-				loanContract.minCratio(),
-			]).catch((err) => {
-				console.log('getLoanIndices', err);
-				throw err;
-			});
-			const loanIndices = [];
-			for (let i = 0; i < n; i++) {
-				loanIndices.push(i);
-			}
-			return { type, minCRatio: wei(minCRatio), loanIndices };
-		};
-
-		const getLoans = async ({ type, minCRatio, loanIndices }: Record<any, any>) => {
-			return Promise.all(loanIndices.map(getLoan.bind(null, type, minCRatio)));
-		};
-
-		const getLoan = async (type: string, minCRatio: Wei, loanIndex: number) => {
-			const loanStateContract = loanStateContracts[type];
-			return {
-				type,
-				minCRatio,
-				loan: await loanStateContract.loans(address, loanIndex),
-			};
-		};
-
-		const makeLoan = async ({ loan, type, minCRatio }: Record<any, any>) => {
-			const loanContract = loanContracts[type];
-			return {
-				...loan,
-				type,
-				minCRatio,
-				cratio: await loanContract.collateralRatio(loan),
-				collateralAsset: COLLATERAL_ASSETS[type],
-				debtAsset: SYNTH_BY_CURRENCY_KEY[loan.currency],
-			};
-		};
-
 		// subscribe to loan open+close
 		const subscribe = () => {
-			for (const type in loanContracts) {
-				const loanContract = loanContracts[type];
-				const loanStateContract = loanStateContracts[type];
-
-				const fetchLoan = async (owner: string, id: string) => {
-					return makeLoan({
-						loan: await loanStateContract.getLoan(owner, id),
-						type,
-						minCRatio: await loanContract.minCratio(),
+			loanContracts.forEach((loanContract, i) => {
+				if (!loanContract) return;
+				const loanStateContract = loanStateContracts[i];
+				const updateLoan = async (owner: string, id: BigNumber) => {
```

Comment:
maybe a `Promise.all` is better here

Reply:
That big `forEach` is just setting up events and calling appropriate functions when the events happening. No async stuff happening